google/ads/google_ads/client.py

```diff
@@ -323,17 +331,184 @@ def intercept_unary_unary(self, continuation, client_call_details, request):
+            metadata_json: request metadata (i.e. headers) in JSON form
+            request_id: unique ID for the request provided in the response
+            request_json: the request object in JSON form
+            trailing_metadata_json: metadata from the response as JSON
+            response_json: the response object as JSON
+        """
+        _logger.debug(self._FULL_REQUEST_LOG_LINE
+                      % (
+                          method,
+                          self.endpoint,
+                          metadata_json,
+                          request_json,
+                          trailing_metadata_json,
+                          response_json
+                      ))
+
+        _logger.info(self._SUMMARY_LOG_LINE
+                     % (
+                         customer_id,
+                         self.endpoint,
+                         method,
+                         request_id,
+                         False,
+                         None
+                     ))
+
+    def _log_failed_request(self, method, customer_id, metadata_json,
+                            request_id, request_json,
+                            trailing_metadata_json, response_json,
+                            fault_message):
+        """Handles logging of a failed request
+
+        Args:
+            method: the gRPC method of the request
+            customer_id: the customer ID associated with the request
+            metadata_json: request metadata (i.e. headers) in JSON form
+            request_id: unique ID for the request provided in the response
+            request_json: the request object in JSON form
+            trailing_metadata_json: metadata from the response as JSON
+            response_json: the response object as JSON
+            fault_message: the error message from the failed request
+        """
+        _logger.warning(self._SUMMARY_LOG_LINE
+                        % (
+                            customer_id,
+                            self.endpoint,
+                            method,
+                            request_id,
+                            True,
+                            fault_message
+                        ))
+
+        _logger.info(self._FULL_FAULT_LOG_LINE
+                     % (
+                         method,
+                         self.endpoint,
+                         metadata_json,
+                         request_json,
+                         trailing_metadata_json,
+                         response_json
+                     ))
+
+    def _log_request(self, client_call_details, request, response, exception):
+        """Handles logging all requests
+
+        Args:
+            client_call_details: information about the client call
+            request: an instance of a gRPC request
+            response: a gRPC response object
+            exception: a gRPC exception object
+        """
+        method = client_call_details.method
```

Comment:
Will we do all of this even if logging is disabled? A minor performance enhancement may be to only do these if we actually need them.

Reply:
We probably should not - I'll bundle this with your previous comment (https://github.com/googleads/google-ads-python/pull/20#discussion_r245814328) and run a couple quick checks before executing all of this. Great suggestion :+1:

Reply:
Moving this comment here:

I updated this so that the logging functionality only fires if the logger is configured to log at `WARNING` or below. See these two commits:

5e25a651ff437f5bbe4b8638a38a412cbb2509ed
5f6abff1c27a5addde89a346427622e21d01e1a6